[Docs] Translate visualization.md (#206)
open-mmlab/mmyolo: docs/en/user_guides/visualization.md

# Visualization

## Feature map visualization

<div align=center>
<img src="https://user-images.githubusercontent.com/89863442/190903635-27bbc619-9bf8-43a8-aea8-ea13b9dad28c.jpg" width="1000" alt="image"/>
</div>

> Visualization can provide an intuitive explanation to the training and testing process of the deep learning model.

Reviewer: ```suggestion
Visualization can provide an intuitive explanation of the training and testing process of the deep learning model.
```

> You can use the `Visualizer` provided in MMYOLO for feature map visualization, which has the following features:

Reviewer 2: 在 MMYOLO 中使用 MMEngine 的可视化器，不是在 MMYOLO 中

Author: > 在 MMYOLO 中使用 MMEngine 的可视化器，不是在 MMYOLO 中

Changed to 'In MMYOLO, you can use the `Visualizer` provided in MMEngine ...'

- Support basic drawing interfaces and feature map visualization.
- Support selecting different layers in the model to get the feature map. The display methods include `squeeze_mean`, `select_max`, and `topk`. Users can also customize the layout of the feature map display with `arrangement`.

## Feature map generation

You can use `demo/featmap_vis_demo.py` to get a quick view of the visualization results. To better understand all functions, we list all primary parameters and their features here as follows:

> - `img`: Identify the image to visualize. Either a single image file or a list of image paths is supported.

Reviewer 2: Identify 不需要

Author: > Identify 不需要

Removed all the 'identify'

> - `img`: used to identify the image to visualize. Either a single image file or a list of image paths is supported.
> 
> - `config`: used to identify the configuration file for the algorithm.
> 
> - `checkpoint`: used to identify the weight file of the corresponding algorithm.
> 
> - `--out-file`: used to identify the path and file name to save the obtained feature map on your device.
> 
> - `--device`: used to identify the hardware used for image inference. For example, `--device cuda:0` means use the first GPU, whereas `--device cpu` means use CPU.
> 
> - `--score-thr`: used to identify the confidence threshold. Only bboxes whose confidence scores are higher than this threshold will be displayed.
> 
> - `--preview-model`: used to identify if there is a need to preview the model. This could make users understand the structure of the feature layer more straightforwardly.
> 
> - `--target-layers`: used to identify the specific layer to get its visualized feature map.
> 
>   - when there is only one parameter, the feature map of that specific layer will be visualized. For example, `--target-layers backbone` ,  `--target-layers neck` ,  `--target-layers backbone.stage4`, etc.
>   - when the parameter is a list, all feature maps of the corresponding layers will be visualized. For example, `--target-layers backbone.stage4 neck` means that the stage4 layer of the backbone and the three layers of the neck are output simultaneously, a total of four layers of feature maps.

Reviewer 3: ```suggestion
- `img`: Identify the image to visualize. Either a single image file or a list of image paths is supported.

- `config`: Identify the configuration file for the algorithm.

- `checkpoint`: Identify the weight file of the corresponding algorithm.

- `--out-file`: Identify the path and file name to save the obtained feature map on your device.

- `--device`: Identify the hardware used for image inference. For example, `--device cuda:0` means use the first GPU, whereas `--device cpu` means use CPU.

- `--score-thr`: Identify the confidence threshold. Only bboxes whose confidence scores are higher than this threshold will be displayed.

- `--preview-model`: Identify if there is a need to preview the model. This could make users understand the structure of the feature layer more straightforwardly.

- `--target-layers`: Identify the specific layer to get its visualized feature map.

  - When there is only one parameter, the feature map of that specific layer will be visualized. For example, `--target-layers backbone` ,  `--target-layers neck` ,  `--target-layers backbone.stage4`, etc.
  - When the parameter is a list, all feature maps of the corresponding layers will be visualized. For example, `--target-layers backbone.stage4 neck` means that the stage4 layer of the backbone and the three layers of the neck are output simultaneously, a total of four layers of feature maps.
```

- `--channel-reduction`: used to identify if there is a need to compress multiple channels into a single channel and then display it overlaid with the picture as the input tensor usually has multiple channels. Three parameters can be used here:

  - `squeeze_mean`: The input channel C will be compressed into one channel using the mean function, and the output dimension becomes (1, H, W).
  - `select_max`: Sum the input channel C in the spatial space, and the dimension becomes (C, ). Then select the channel with the largest value.
  - `None`: Indicates that no compression is required. In this case, the `topk` feature maps with the highest activation degree can be selected to display through the `topk` parameter.

> - `--topk`: Valid only when the `channel_reduction` parameter is `None`. Once it is valid, it selects the `topk` channels according to the activation degree, and then display i

Reviewer: Missing some words

Author: > Missing some words

Resolved.

> - `--topk`: Valid only when the `channel_reduction` parameter is `None`. Once it is valid, it selects the `topk` channels according to the activation degree, and then display i

Reviewer: ```suggestion
- `--topk`: Valid only when the `channel_reduction` parameter is `None`. Once it is valid, it selects the `topk` channels according to the activation degree, and then displays i
```

> - `--topk`: Valid only when the `channel_reduction` parameter is `None`. It selects the `topk` channels according to the activation degree and then displays it overlaid with the image. The display layout can be specified using the `--arrangement` parameter, which is a array of two numbers separated with space. For example, `--topk 5 --arrangement 2 3` means the five feature maps with the highest activation degree are displayed in `2 rows and 3 columns`. Similarly, `--topk 7 --arrangement 3 3` means the seven feature maps with the highest activation degree are displayed in `3 rows and 3 columns`.

Reviewer: ```suggestion
- `--topk`: Valid only when the `channel_reduction` parameter is `None`. It selects the `topk` channels according to the activation degree and then displays it overlaid with the image. The display layout can be specified using the `--arrangement` parameter, which is an array of two numbers separated by space. For example, `--topk 5 --arrangement 2 3` means the five feature maps with the highest activation degree are displayed in `2 rows and 3 columns`. Similarly, `--topk 7 --arrangement 3 3` means the seven feature maps with the highest activation degree are displayed in `3 rows and 3 columns`.
```

  - If `topk` is not -1, topk channels will be selected to display in order of the activation degree.
  - If `topk` is -1, channel number C must be either 1 or 3 to indicate that the input data is a picture. Otherwise, an error will prompt the user to compress the channel with `channel_reduction`.

- Considering that the input feature map is usually very small, the function will upsample the feature map by default for easy visualization.

> **Note: When the image and feature map scales are different, the `draw_featmap` function will automatically perform upsampling alignment. If your image has an operation such as `Pad` in the preprocessing during the inference, the feature map obtained is processed with `Pad`, which may cause misalignment problems if you directly upsample the image.**

Reviewer: ```suggestion
**Note: When the image and feature map scales are different, the `draw_featmap` function will automatically perform an upsampling alignment. If your image has an operation such as `Pad` in the preprocessing during the inference, the feature map obtained is processed with `Pad`, which may cause misalignment problems if you directly upsample the image.**
```
